pkg/injector/bootstrap.go

```diff
@@ -101,3 +105,118 @@ func (wh *mutatingWebhook) marshalAndSaveBootstrap(name, namespace string, confi
 	log.Debug().Msgf("Creating bootstrap config for Envoy: name=%s, namespace=%s", name, namespace)
 	return wh.kubeClient.CoreV1().Secrets(namespace).Create(context.Background(), secret, metav1.CreateOptions{})
 }
+
+// NewBootstrapSecretRotator returns a new bootstrap secret rotator.
+func NewBootstrapSecretRotator(kubeController k8s.Controller, certManager *certificate.Manager, checkInterval time.Duration) *BootstrapSecretRotator {
+	return &BootstrapSecretRotator{
+		kubeController: kubeController,
+		certManager:    certManager,
+		checkInterval:  checkInterval,
+	}
+}
+
+// getBootstrapSecrets returns the bootstrap secrets stored in the informerCollection's store.
+func (b *BootstrapSecretRotator) getBootstrapSecrets() []*corev1.Secret {
+	secrets := b.kubeController.ListSecrets()
+	var bootstrapSecrets []*corev1.Secret
+
+	for _, secret := range secrets {
+		// finds bootstrap secrets
+		if strings.Contains(secret.Name, bootstrapSecretPrefix) {
+			bootstrapSecrets = append(bootstrapSecrets, secret)
+		}
+	}
+	return bootstrapSecrets
+}
+
+// rotateBootstrapSecrets updates the bootstrap secret by getting the current or issuing a new certificate.
+func (b *BootstrapSecretRotator) rotateBootstrapSecrets(ctx context.Context) {
+	bootstrapSecrets := b.getBootstrapSecrets()
+	for _, secret := range bootstrapSecrets {
+		bootstrapCert, err := getCertFromSecret(secret)
+		if err != nil {
+			log.Error().Err(err).Str(errcode.Kind, errcode.GetErrCodeWithMetric(errcode.ErrObtainingCertFromSecret)).
+				Msgf("Error getting cert %s from secret %s/%s", bootstrapCert, secret.Namespace, secret.Name)
+		}
+
+		if !b.certManager.ShouldRotate(bootstrapCert) {
+			continue
+		}
+
+		issuedCert, err := b.certManager.IssueCertificate(bootstrapCert.CommonName.String(), certificate.Internal, certificate.FullCNProvided())
+		if err != nil {
+			log.Error().Err(err).Str(errcode.Kind, errcode.GetErrCodeWithMetric(errcode.ErrRotatingCert)).Msgf("Error rotating cert %s", issuedCert)
+		}
+
+		secret.Data[bootstrap.EnvoyXDSCACertFile] = issuedCert.GetTrustedCAs()
+		secret.Data[bootstrap.EnvoyXDSCertFile] = issuedCert.GetCertificateChain()
+		secret.Data[bootstrap.EnvoyXDSKeyFile] = issuedCert.GetPrivateKey()
+
+		err = b.kubeController.UpdateSecretData(ctx, secret, secret.Data)
+		if err != nil {
+			if apierrors.IsConflict(err) {
+				log.Error().Err(err).Str(errcode.Kind, errcode.GetErrCodeWithMetric(errcode.ErrUpdatingBootstrapSecret)).
```

Comment:
is this more of a warning?  Is there enough wiggle room in expiration in `ShouldRotate()` to absorb this error? I.e.  If we have a conflict and then postpone updating for another round will we get to rotation before we cause a disruption?

 It looks like `30s` for ShouldRotate (https://github.com/openservicemesh/osm/blob/194856e2fbe909cc9d2830951aed97e9e7ddb033/pkg/certificate/certificate.go#L13)  and we run every 5 seconds.  The thing to be wary of would be large cluster or other delays that might cause this to tick up to high.

There is a helper function in k8s libraries that will do retry on conflict that could eliminate any potential for this being a  sneaky bug long term

Reply:
https://github.com/kubernetes/client-go/blob/v0.25.0/util/retry/util.go#L103

Reply:
Docs for `RetryOnConflict`: https://pkg.go.dev/k8s.io/client-go/util/retry#RetryOnConflict